src/etc/inc/pkg-utils.inc

```diff
@@ -359,21 +361,46 @@ function get_pkg_info($pkgs = 'all', $only_local = false) {
 
 	/* Do not run remote operations if pkg has a lock */
 	if (is_subsystem_dirty('pkg')) {
-		$only_local = true;
+		$local_only = true;
 		$lock = false;
 	} else {
 		$lock = true;
 	}
 
 	$extra_param = "";
-	if ($only_local) {
+	if ($local_only) {
 		$extra_param = "-U ";
 	}
 
 	if ($lock) {
 		mark_subsystem_dirty('pkg');
 	}
-	$rc = pkg_exec("search {$extra_param}-R --raw-format json-compact " . $pkgs, $out, $err);
+
+	if (!$installed_only) {
+		// repo catalog search (either remote or local_only)
+		$rc = pkg_exec("search {$extra_param}-R --raw-format json-compact " . $pkgs, $out, $err);
+	}
+	if ($installed_only || ($local_only && $rc != 0)) {
+		/* use pkg info if (1) installed pkg search or (2) local catalog copy search requested + failed.
+		 *
+		 * The local repo catalog copy may be cleared if a previous call to pkg search couldn't get the
+		 * remote repo catalog.
+		 *
+		 * If the calling code would have accepted local copy info (which isn't assumed up to date) then it
+		 * makes sense to fall back on pkg info to at least return the known info about installed pkgs (pkg
+		 * info should still work), instead of failing and returning no info at all. For example, this
+		 * enables offline view + management of installed pkgs.
+		 */
+
+		// pkg info errors if none match, unlike pkg search, so test it will work beforehand...
+		$tmp1 = $tmp2 = '';
+		$tmprc = pkg_exec("info -e -R --raw-format json-compact " . $pkgs, $tmp1, $tmp2);
```

Comment:
Thanks, however I think here it's best not to do that.

The aim in this snippet is very specific - to identify in advance whether `pkg info` will return a "no matching packages" error for the exact parameters in `$pkgs`.  It's not _per se_ trying to determine if a package(s) are installed, rather if, given the specific args in the code, will an error be returned due to no match.

Unless is_pkg_installed() returns exactly the same as `pkg info` for the given arg string, it can't be used here, and if it's doing exactly the same then it's probably wrapping pkg info and a direct call to pkg info here will be more direct and add clarity.

tl;dr as we aren't trying to detect whether packages are installed, and we're trying to detect if pkg info will return an error specifically due to no matches on arg string, the best approach is to use pkg info itself to do that test, not a wrapper or other internal function.